File: csharp/exercises/Webshop.Service/Payment/PaymentService.cs
using Webshop.Service.Common;
using Webshop.Service.Users;

namespace Webshop.Service.Payment;

public interface IPaymentService
{
    void HandlePayment(string userId, decimal amountInNok);
}

public class PaymentService : IPaymentService
{
    private readonly ThirdPartyPaymentIntegration _thirdPartyPaymentIntegration;

    public PaymentService()
    {
        _thirdPartyPaymentIntegration = new ThirdPartyPaymentIntegration();
    }

    public void HandlePayment(string userId, decimal amountInNok)
    {
        var userService = new UserService(new UserRepository(), new PasswordValidator());
        var user = userService.GetUser(userId);
        if (user == null)
        {
            Logger.LogError($"Attempted to handle payment for non-existing user {userId}");
            throw new Exception($"Could not find user with id {userId}");
        }

        if (amountInNok < 0)
            throw new Exception("Amount cannot be negative");

        var paymentOk = _thirdPartyPaymentIntegration.HandlePaymentForUser(userId, amountInNok, "WebShop payment");
        if (!paymentOk)
        {
            Logger.LogError($"Payment failed for user {userId} with amount {amountInNok}");
            throw new Exception("Payment failed");
        }

        var notificationService = new NotificationService();
        notificationService.SendMail(user.Email, "Payment succeeded");
    }
}

File: java/solution/src/main/java/Webshop/Service/Payment/PaymentService.java
package Webshop.Service.Payment;

import Webshop.Service.Common.Logger;
import Webshop.Service.Common.NotificationService;
import Webshop.Service.Users.UserRepository;
import Webshop.Service.Users.UserService;
import Webshop.Service.Common.PasswordValidator;
import Webshop.Service.Users.UserRepository;
import Webshop.Service.Users.User;

interface IPaymentService {
    void handlePayment(String userId, double amountInNok);
}

public class PaymentService implements IPaymentService {
    private final ThirdPartyPaymentIntegration thirdPartyPaymentIntegration;

    public PaymentService() {
        this.thirdPartyPaymentIntegration = new ThirdPartyPaymentIntegration();
    }

    @Override
    public void handlePayment(String userId, double amountInNok) {
        UserService userService = new UserService(new UserRepository(), new PasswordValidator());
        User user = userService.getUser(userId);
        if (user == null) {
            Logger.logError("Attempted to handle payment for non-existing user " + userId);
            throw new RuntimeException("Could not find user with id " + userId);
        }

        if (amountInNok < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }

        boolean paymentOk = thirdPartyPaymentIntegration.handlePaymentForUser(userId, amountInNok, "WebShop payment");
        if (!paymentOk) {
            Logger.logError("Payment failed for user " + userId + " with amount " + amountInNok);
            throw new RuntimeException("Payment failed");
        }

        NotificationService notificationService = new NotificationService();
        notificationService.sendMail(user.getEmail(), "Payment succeeded");
    }
}

